### ISSUE TITLE
Log Sourcegraph extensions events in BigQuery

### ISSUE THREAD

**Issue by ebrodymoore** [sourcegraph/sourcegraph#15007]
Background: there's a huge effort going on right now [build extensions](https://docs.google.com/spreadsheets/d/1Ht97I5n9XC5qG2JWMUMog_8kPL7cdokOGB8yIbu3y1k/edit#gid=0) as well as the [UX around extension discoverability](https://github.com/sourcegraph/sourcegraph/issues/14165). 

We have access to user settings for Cloud users but extensions are not really useful for these users so we've deprioritized this data. Instead, we've learned that almost all extensions usage across all instances (including Cloud) could be tracked. This would be _huge_ to have this data to better inform engagement and extensions strategy. To do this, we need the Cloud team's help 😃 

Context from @dadlerj: 
> The majority of self-hosted customer instances actually pull extensions from Sourcegraph.com/extensions (the clients pull this, so each user’s browser downloads each extension on every pageview), we SHOULD actually already have really good usage/download data… We should check on network access, and perhaps have some sort of backend logging on this. 
> 
> How we log this—as a “backend event” through our event_logs table and into BigQuery, or as some server log that goes into Stackdriver/LogDNA—is up to the cloud team (and perhaps @felixfbecker / @ebrodymoore ). The volume of these pulls must be insanely gigantic though, so we should be ready for this to absolutely explode the size of our sourcegraph.com event_logs DB if we use that method… Perhaps we create some custom pipeline to log these in BigQuery without storing them in Postgres as well

**Comment by ebrodymoore** [comment]
@tsenart @christinaforney let me know if you what information to effectively prioritize this. It's obviously not fully fleshed out so we're definitely still in the 'scoping' phase and need the cloud team's input!

**Comment by tsenart** [comment]
@ebrodymoore, @felixfbecker, @Joelkw: Let's define what metrics you want to have first, and work on a technical solution from those — it may be based on event_logs or not. What are the metrics you want and why?

**Comment by felixfbecker** [comment]
@tsenart please see https://docs.google.com/document/d/1NHi0nDohuPLdaJjV0l-Bk7NG_pI-A8s0H5CSzE9RWrQ/edit

**Comment by ebrodymoore** [comment]
To add to this: [this bookmark](https://docs.google.com/document/d/1NHi0nDohuPLdaJjV0l-Bk7NG_pI-A8s0H5CSzE9RWrQ/edit#bookmark=id.5gvstvdmxakm) is the required data points we'll need to end up getting, and this one is the [nice-to-haves](https://docs.google.com/document/d/1NHi0nDohuPLdaJjV0l-Bk7NG_pI-A8s0H5CSzE9RWrQ/edit#bookmark=id.b0g4bid6n88f)

@Joelkw and I know the doc got a little messy so we're happy to clarify anything and/or be part of the solution!

**Comment by christinaforney** [comment]
@tsenart and @arussellsaw - FYI this issue is blocking the web team's ability to have some important metrics. Just flagging here that if we could get this in before branch cut so it gets into the next release, that would be super helpful!

**Comment by tsenart** [comment]
@arussellsaw: Branch cut is on the 15th, I believe, you can check the shared Sourcegraph calendar to check. I think we have enough time!

**Comment by felixfbecker** [comment]
I may be wrong, but as I understand it these changes would be purely on sourcegraph.com so the branch cut would not be relevant, but having these in `main` and deployed to dotcom as soon as possible would be great (we are blocked on this for measuring our goals for Q3 and Q4).

**Comment by arussellsaw** [comment]
just a thought, does the change to the firefox extension cause issues for this? https://sourcegraph.slack.com/archives/CMT39K56Z/p1604423319147700?thread_ts=1604422281.145500&cid=CMT39K56Z we now bundle the extensions directly into the add-on. If we don't need this data from _all_ users then we still get it from the chrome extension.

or am i mixing up the two concepts, sourcegraph server extensions vs browser extensions? i know the browser extensions load code from the server extensions...

**Comment by felixfbecker** [comment]
We only really care about tracking non-default extensions, i.e. the ones you have to explicitly enable that are not code intel extensions or git extras. What's bundled in Firefox is only the default extensions: code intel extensions and git extras. So basically yes, we wouldn't get any data from Firefox browser extension users, but those users unfortunately also don't contribute to the interesting extension usage (which is a situation we may want to solve somehow eventually).

**Comment by arussellsaw** [comment]
some questions i need to answer from our call: 
* Right now sourcegraph.com is behind a CDN, we need to make sure that it's viable to bypass the CDN for extension loads, which might be expensive data wise, but we may also be able to send a cache freshness check request every time that doesn't send the full payload.
* We'll be using the extension load HTTP request to emit an event, is this scaleable in the long term? we might want to look at an alternate events pipeline for high-frequency events.

I have personal biases on the events pipeline, but i think it would be worthwhile long term for sourcegraph.com to have it's own high-throughput events pipeline as it will be very useful for analytics in the future. Some kind of thing where we can just throw arbitrary events into a pubsub topic with a JSON payload, and have them output to date/time/event type partitioned bigquery tables would be super handy.

as a note i'll implement support for passing an anonymised user ID on the request today, so that we don't need to do anything more when frontend lands on a decision on how to approach these IDs

**Comment by dadlerj** [comment]
>I have personal biases on the events pipeline, but i think it would be worthwhile long term for sourcegraph.com to have it's own high-throughput events pipeline as it will be very useful for analytics in the future. Some kind of thing where we can just throw arbitrary events into a pubsub topic with a JSON payload, and have them output to date/time/event type partitioned bigquery tables would be super handy.

We have this already actually. For the most part, you just described our cloud data pipeline. Events on cloud are published in `publishSourcegraphDotComEvent` at https://sourcegraph.com/github.com/sourcegraph/sourcegraph@88ddadf10773b4d6c877593d4b6c691d112b6e0d/-/blob/internal/usagestats/event_handlers.go#L87:16

We could always provide a flag that _doesn't_ log these particular events in the postgres db to avoid expanding it, so that they only appear in bigquery.

**Comment by tsenart** [comment]
The @sourcegraph/web team is taking care of this, but I think they have a separate GitHub issue for it.
